### ISSUE TITLE
motor commands not working

### ISSUE THREAD

**Issue by alesolano** [rhempel/ev3dev-lang-python#139]
I was testing some motor commands. `run-direct` command and `run_direct()` function work fine. However, `run-to-abs-pos` and `run-timed` commands, along with `run_to_abs_pos()` and `run_timed()` functions,  are not in the mood to work. (They don't throw any errors, they just do nothing).

These are my .py files. Hope I was using the functions properly.

```
#!/usr/bin/python

from ev3dev.auto import *

m = LargeMotor(OUTPUT_A)

m.position_sp = 360
m.run_to_abs_pos()
```

```
#!/usr/bin/python

from ev3dev.auto import *

m = LargeMotor(OUTPUT_A)

m.time_sp = 1000
m.run_timed()
```

Also, I don't know how to use position_p attribute. It always comes with this error.

```
>>> m.position_p
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/ev3dev/core.py", line 416, in position_p
    return self.get_attr_int('hold_pid/Kp')
  File "/usr/lib/python2.7/dist-packages/ev3dev/core.py", line 207, in get_attr_int
    return int(self._get_attribute(attribute))
  File "/usr/lib/python2.7/dist-packages/ev3dev/core.py", line 200, in _get_attribute
    return self._attribute_cache.read(abspath(self._path + '/' + attribute))
  File "/usr/lib/python2.7/dist-packages/ev3dev/core.py", line 85, in read
    return f.read().decode().strip()
IOError: [Errno 95] Operation not supported
```

Thank you!

**Comment by ddemidov** [comment]
You need to specify the amount of power to apply to the motor. When `m.speed_regulation_enabled` is `'off'` this is done with `duty_cycle_sp` attribute (the acceptable range is [-100, 100]):
```py
m.run_to_rel_pos(speed_regulation_enabled='off', duty_cycle_sp=100, position_sp=360)
```
When `speed_regulation_enabled` is `'on'`, you need to use `speed_sp` instead of `duty_cycle_sp`:
```py
m.run_to_rel_pos(speed_regulation_enabled='on', speed_sp=900, position_sp=360)
```
The acceptable range in this case depends on the motor type. See http://www.ev3dev.org/docs/tutorials/tacho-motors/ for more details.

Note that this will change with the next kernel release; only `run-direct` command will accept `duty_cycle_sp`, and the rest of the commands will always work with `speed_sp`.

`position_p` is one of PID parameters for holding the motor in place (when `stop_command` is `hold`), and is not supported for tacho motors.

**Comment by alesolano** [comment]
Thank you! 

I've been trying to turn a Large Motor 360 degrees. I have come with 3 options:

```
#!/usr/bin/python

from ev3dev.auto import *

m = LargeMotor(OUTPUT_A)
m.speed_regulation_enabled='off'
m.duty_cycle_sp = 60
m.position = 0

b = Button()

m.position_sp = 360

#Option 1
#m.run_to_abs_pos()

#Option 2
#while not m.position == 360:
#    m.run_to_abs_pos()

#Option 3
while not b.any():
    m.run_to_abs_pos()
```

And only the Option 3 seems to work, although showing a lot of wobble. The Option 1 don't reach the required position and the Option 2 don't stop the motor at time.

What is the best way to do this?

**Comment by ddemidov** [comment]
The commands are asynchronous, so you need to wait for the motor to stop turning. For example, like this:
```py
while 'running' in m.state: time.sleep(0.01)
```

Right now your script exits immediately after sending the command to the motor, and the motor resets (stops) on script exit.

**Comment by alesolano** [comment]
It works perfect! Thanks a lot! I procede to close the issue

**Comment by alesolano** [comment]
Hey Denis!
I'm facing the same problem, but now for two motors at the same time.I don't know where to put the line 
```python
while 'running' in m.state: time.sleep(0.01)
```
if I'm handling with two motors.

I explain the case better [here](https://github.com/alesolano/ev3-mapping/issues/2).


Thank you!!

**Comment by ddemidov** [comment]
I think there are a couple of options:

* If you know exactly how far each of the wheel should turn, then you start both of them and then wait for both of them to stop turning. This can be done either with:
```py
while 'running' in left.state and 'running' in right.state: sleep(0.01)
```
or with
```py
while 'running' in left.state: sleep(0.01)
while 'running' in right.state: sleep(0.01)
```
The latter seems to be more effective, since you query single motor in the first loop, and then make sure the second motor also stopped with the second loop.

* If you know how far one of the motors should turn (for example, the one on the outer arc of the turn), then you can start that motor with `run-to-[rel,abs]-pos`, start the other with `run-forever`, wait for the first one to stop, and then immediately stop the other.

I dont see anything else until ev3dev/ev3dev#454 is solved.

**Comment by Fuzzyma** [comment]
I use the nightly build of the 30.12.2015. `run_timed` and `run_forever` are not working. I followed the examples which state this code:

```py
m.run_timed(time_sp=3000, duty_cycle_sp=75)
```

However: e.g. `run_direct` works

**Comment by WasabiFan** [comment]
The `run_timed` and `run_forever` commands run the motor at a specific speed, not with a duty cycle. So you need to set `speed_sp` instead of `duty_cycle_sp`. See ev3dev/ev3dev#715. What examples are you referring to?

**Comment by Fuzzyma** [comment]
http://ev3dev-lang-python.readthedocs.io/en/latest/#example-code

I see the word `latest` in the url. So I thought it would match my needs :D

**Comment by WasabiFan** [comment]
That is fixed in the main README in the develop branch, but the changes haven't been published to the docs site yet. I imagine that's because this module's mot recent version is still technically a "release candidate" so we haven't tagged a GitHub release or published the docs.

@dwalton76 @ddemidov We really need to figure out what needs to be done for a real release, make those changes, and then publish.

**Comment by dwalton76** [comment]
My guess is in the past the releases here were coordinated with the release of a new ev3dev binary?  I do not know when the next scheduled binary is but it has been a while (December).  Tagging @dlech

**Comment by dlech** [comment]
Past releases were totally uncoordinated. :wink:

IIRC, this repository is setup so that if you create an annotated tag, PyPi will pick up the release. Note that an annotated tag is not the same as `git tag xyz` - you need to add an extra option for the annotation. Not sure what it is off of the top of my head.

If you want the debian package in ev3dev to match what is on PyPi, then we need to build it from the same tag. If someone wants to learn how to do this, let me know and I will set you up so you can push packages to `archive.ev3dev.org`, otherwise, I can take care of it.

We can do a version release of ev3dev-lang-python anytime you feel that it is in a releasable state. There is no need to wait for an ev3dev image release.

**Comment by WasabiFan** [comment]
@dwalton76 Thanks a lot for taking the time to do an issue clean-up; this makes it much easier to see what needs to be done!

I don't see any ship-stopping bugs ATM, so right now I think the only thing on my mind that we should be waiting for is updating those TODOs in the README with links to the page(s) that are currently in PR on the main site. Once that happens (the requisite pages are live), we should be good-to-go. Is there anything else that you see that we need to do?

**Comment by dwalton76** [comment]
Once those few TODOs are knocked out I can't think of anything else to hold us back from doing a release. 

@dlech if you have some notes on how the packages where built in the past I can run with it.

**Comment by dlech** [comment]
http://www.ev3dev.org/docs/devtools/packaging-for-ev3dev/

**Comment by jrayers** [comment]
Hello, I believe my problem is related to this issue in some way.  Tonight I started from square one by following the docs here:

http://www.ev3dev.org/docs/getting-started/

Using that I am successfully able to SSH into my ev3 brick via Putty.  The brick is connected to my Win10 pro PC via USB.  After that I am manually running python3 from $ in Putty.  My goal at this point is to manually complete the 3 usage examples listed on GitHub, here:

https://github.com/rhempel/ev3dev-lang-python

From the examples listed in the link above, I have no problem importing the library, doing 'Controlling the LEDs with a touch sensor' and even 'Using text-to-speech,' so I feel pretty good about things being set up correctly.  However, I cannot complete the 'Running a motor' example.  It simply does nothing.

At the Python prompt, I am typing the following per the example:

`>>> m = ev3.LargeMotor('outA')`
`>>> m.run_timed(time_sp=3000, speed_sp=500)`

Nothing happens at all.  The motor really does seem set up correctly despite not running, for example:

1. On the ev3 brick in the 'Device Browser/Motors' section I see 'lego-ev3-l-motor at outA'
2. If I manually turn the motor some, then type:

`>>> m.position`
`976`

(turn motor manually again some)

`>>> m.position`
`895`

Given that it returns positions, that seems good.  I am just not sure where else to go at this point to make this motor turn.  I have confirmed it works fine from the normal ev3 software environment, so I must be missing something key along the way.

More relevant info from the ev3 screen, running: brickman v0.7.2 with kernel: 3.16.7-ck21-9-ev3dev-ev3 and revision: 0006

Thanks a lot for all this, I really appreciate your efforts, and please tell me if I should create a new issue or go to another one.

**Comment by ddemidov** [comment]
Hi,

The last released image (that is pointed at in the Getting Started page) is outdated, which means that the motor API described in all of the documentation is not the one that is actually used in the `9-ev3dev` kernel that you have. I think the easiest way to use the image you have (without upgrading anything)  would be to add `speed_regulation_enabled='on'` to all motor commands (as in `m.run_timed(time_sp=3000, speed_regulation_enabled='on', speed_sp=500)`.

Another option would be to reflash your SD card with a nightly image. That is currently the easiest and the fastest way to get the latest kernel and the latest version of the python library.

**Comment by WasabiFan** [comment]
The version of the README displayed by default is for a newer version of the library than that which you are currently using (we just released a new version of this library two days ago). We're still waiting for an ev3dev image release to include the newer library.

For the version that you are using, you should add "speed_regulation='on'" to the parameters when calling that function to make it work.

Once we release a new image (hopefully in the next few days), you will probably want to re-flash your card with it to get the newest version of the library and operating system.

@dlech are we currently blocked by anything for a release?

**Comment by WasabiFan** [comment]
Whoops, @ddemidov and I responded at the same time :laughing:

**Comment by ddemidov** [comment]
Just checked, its `speed_regulation_enabled`: https://github.com/rhempel/ev3dev-lang-python/blob/0.6.0/ev3dev/core.py#L511

**Comment by WasabiFan** [comment]
That must've been a late-night naming decision on my part... 😞

**Comment by jrayers** [comment]
Thanks a TON for the help.  I figured it had to be something fairly straightforward given how much of my setup was working.  I will post back with results, likely will not be until tomorrow at this point.

Thanks again.

**Comment by WasabiFan** [comment]
No problem! If you'd like to get ahead of the curve before an official image is released (which means all the documentation would actually refer to the versions you are using), you can re-flash your SD card with the latest nightly build: https://oss.jfrog.org/list/oss-snapshot-local/org/ev3dev/brickstrap/2016-09-30/. Then you wouldn't need to change your code once we finally do release a new official image, and you can take advantage of the latest features and bugfixes. If you want to stick with what you have, that's fine too.

**Comment by WasabiFan** [comment]
Ah, I didn't realize that we were using milestones for absolute work tracking now. Good to know!

**Comment by jrayers** [comment]
For whatever reason inserting the `"speed_regulation='on'"` did not work for me.  Not sure why, the issue could lie between the keyboard and the chair.  Anyway, I re-flashed to the nightly and all is well.  Thanks again.
